### PR TITLE
experimental: support breakpoints and states in style object model

### PR DESCRIPTION
**Pull request by TrySound**
Ref https://github.com/webstudio-is/webstudio/issues/1536

Here improved sorting declared values and computing cascade value. Added support for breakpoints and states. Introduced specificity rules to sort by State-Breakpoint-StyleSource without complicated logic.

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

**Review by kof — APPROVED** [review]

File: apps/builder/app/shared/style-object-model.test.ts [hunk 1]
```diff
@@ -558,3 +588,252 @@ test("resolve non-cyclic references in custom properties", () => {
...
+    },
+  ];
+  const model: StyleObjectModel = {
+    styleSourcesByInstanceId: new Map([["body", ["bodyLocal"]]]),
+    styleByStyleSourceId: getStyleByStyleSourceId(styles),
+  };
+  const styleSelector: StyleSelector = {
+    instanceSelector: ["body"],
+    // large is not active breakpoint
+    activeBreakpoints: ["base", "small"],
```
Comment (kof) [inline apps/builder/app/shared/style-object-model.test.ts:661]: I think the right term is "applicable", e.g. applicableBreakpoints

File: apps/builder/app/shared/style-object-model.test.ts [hunk 2]
```diff
@@ -558,3 +588,252 @@ test("resolve non-cyclic references in custom properties", () => {
...
+    activeStates: new Set(),
```
Comment (kof) [inline apps/builder/app/shared/style-object-model.test.ts:662]: same for states

File: apps/builder/app/shared/style-object-model.test.ts (hunk 1 above)
Comment (TrySound) [reply apps/builder/app/shared/style-object-model.test.ts:661]: This is more about whether provided list of breakpoints is activated from ui and rendered in canvas

Comment (kof) [reply apps/builder/app/shared/style-object-model.test.ts:661]: checked with the spec, that's the term they use

Comment (kof) [reply apps/builder/app/shared/style-object-model.test.ts:661]: > This is more about whether provided list of breakpoints is activated from ui and rendered in canvas

From builder standpoint yes, but from the logic that's the list of breakpoints you need to consider when calculating the value, right? 

I mean these breakpoints conditions are satisfying the viewport potentially, right?

Comment (TrySound) [reply apps/builder/app/shared/style-object-model.test.ts:661]: > checked with the spec, that's the term they use

please attach a link

Comment (kof) [reply apps/builder/app/shared/style-object-model.test.ts:661]: https://www.w3.org/TR/mediaqueries-5/ you will find the word "apply"  and "applicable"

Comment (TrySound) [reply apps/builder/app/shared/style-object-model.test.ts:661]: Very vague. Styles are applicable with certain condition condition. Conditions are not applicable, they define when to apply

Comment (TrySound) [reply apps/builder/app/shared/style-object-model.test.ts:661]: The query already applied/activated/matched

Comment (kof) [reply apps/builder/app/shared/style-object-model.test.ts:661]: then seems it is applicable :)

Comment (TrySound) [reply apps/builder/app/shared/style-object-model.test.ts:661]: applicable - can be potentially applied

here the query already active/matching

Comment (kof) [reply apps/builder/app/shared/style-object-model.test.ts:661]: I think applicable is when they are actually machtching, same semantics, but feel free to use "matching" ... active is just overused and often has too many meanings

**Pull request merged**